src/openrct2/network/NetworkGroups.cpp

```diff
@@ -0,0 +1,304 @@
+    if (platform_file_exists(path))
+    {
+        try
+        {
+            json = Json::ReadFromFile(path);
+        }
+        catch (const std::exception& e)
+        {
+            log_error("Failed to read %s as JSON. Setting default groups. %s", path, e.what());
+        }
+    }
+
+    if (json == nullptr)
+        return true;
+
+    auto addOrUpdateGroup = [&](json_t* jsonGroup) -> void {
+        auto tmpGroup = NetworkGroup::FromJson(jsonGroup);
+
+        // Name always has priority over id.
+        auto* group = GetByName(tmpGroup.GetName());
+        if (group != nullptr && group->IsImmutable())
+        {
+            // Don't modify immutable groups, old group.json files still contain immutable groups.
+            return;
+        }
+
+        // If it doesn't exist, create one with the given name.
+        if (group == nullptr)
+            group = Create(tmpGroup.GetName(), false, true);
+
+        // Update.
+        group->ActionsAllowed = tmpGroup.ActionsAllowed;
+    };
+
+    json_t* json_groups = json_object_get(json, "groups");
+    size_t groupCount = (size_t)json_array_size(json_groups);
+    for (size_t i = 0; i < groupCount; i++)
+    {
+        json_t* jsonGroup = json_array_get(json_groups, i);
+        addOrUpdateGroup(jsonGroup);
+    }
+
+    json_t* jsonDefaultGroup = json_object_get(json, "default_group");
+    NetworkGroupId_t defaultId = static_cast<NetworkGroupId_t>(json_integer_value(jsonDefaultGroup));
+    if (auto* group = GetById(defaultId))
+    {
+        SetDefault(group);
+    }
+
+    json_decref(json);
+
+    return true;
+}
+
+bool NetworkGroups::Save()
+{
+    utf8 path[MAX_PATH];
+
+    platform_get_user_directory(path, nullptr, sizeof(path));
+    safe_strcat_path(path, "groups.json", sizeof(path));
+
+    json_t* jsonGroupsCfg = json_object();
+    json_t* jsonGroups = json_array();
+
+    auto groups = GetAll(true);
+    for (auto* group : groups)
+    {
+        json_array_append_new(jsonGroups, group->ToJson());
+    }
+    json_object_set_new(jsonGroupsCfg, "default_group", json_integer(_defaultId));
+    json_object_set_new(jsonGroupsCfg, "groups", jsonGroups);
+    try
+    {
+        Json::WriteToFile(path, jsonGroupsCfg, JSON_INDENT(4) | JSON_PRESERVE_ORDER);
+    }
+    catch (const std::exception& ex)
+    {
+        log_error("Unable to save %s: %s", path, ex.what());
+    }
+
+    json_decref(jsonGroupsCfg);
+
+    return true;
+}
+
+NetworkGroup* NetworkGroups::GetById(NetworkGroupId_t id) const
+{
+    if (static_cast<size_t>(id) >= _groups.size())
+        return nullptr;
+
+    return _groups[id].get();
+}
+
+NetworkGroup* NetworkGroups::GetByName(const std::string& name) const
+{
+    for (auto& group : _groups)
+    {
+        if (group == nullptr)
+            continue;
+        if (group->GetName() == name)
+            return group.get();
+    }
+    return nullptr;
+}
+
+std::vector<NetworkGroup*> NetworkGroups::GetAll(bool excludeHost) const
```

Comment:
I suggest we make this by default exclude host and have a constexpr value of false that includes the host.

Reply:
I wanted to have this explicit as for network serialization the host must be included. For the UI part its excluded. If it defaults to something the person who is tempted to call this might not be aware what it would do otherwise.